core/trino-main/src/main/java/io/trino/sql/planner/planprinter/HashingAnonymizer.java

```diff
@@ -0,0 +1,355 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.planprinter;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.connector.CatalogName;
+import io.trino.execution.TableInfo;
+import io.trino.metadata.IndexHandle;
+import io.trino.metadata.QualifiedObjectName;
+import io.trino.metadata.TableExecuteHandle;
+import io.trino.metadata.TableHandle;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.ConnectorPartitioningHandle;
+import io.trino.spi.type.Type;
+import io.trino.sql.planner.PartitioningHandle;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.SystemPartitioningHandle;
+import io.trino.sql.tree.BinaryLiteral;
+import io.trino.sql.tree.BooleanLiteral;
+import io.trino.sql.tree.CharLiteral;
+import io.trino.sql.tree.DecimalLiteral;
+import io.trino.sql.tree.DoubleLiteral;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionRewriter;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.GenericLiteral;
+import io.trino.sql.tree.IntervalLiteral;
+import io.trino.sql.tree.Literal;
+import io.trino.sql.tree.LongLiteral;
+import io.trino.sql.tree.NullLiteral;
+import io.trino.sql.tree.StringLiteral;
+import io.trino.sql.tree.SymbolReference;
+import io.trino.sql.tree.TimeLiteral;
+import io.trino.sql.tree.TimestampLiteral;
+
+import java.util.Locale;
+import java.util.Map;
+
+import static com.google.common.hash.Hashing.murmur3_128;
+import static io.trino.sql.planner.Partitioning.ArgumentBinding;
+import static io.trino.sql.planner.plan.StatisticsWriterNode.WriteStatisticsHandle;
+import static io.trino.sql.planner.plan.StatisticsWriterNode.WriteStatisticsTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.CreateTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.DeleteTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.InsertTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.RefreshMaterializedViewTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.TableExecuteTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.UpdateTarget;
+import static io.trino.sql.planner.plan.TableWriterNode.WriterTarget;
+import static io.trino.sql.tree.ExpressionTreeRewriter.rewriteWith;
+import static java.lang.Math.log10;
+import static java.lang.Math.pow;
+import static java.lang.Math.round;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+/**
+ * An implementation of {@link Anonymizer} which uses murmur3 hashing to anonymize the plan
+ */
+public class HashingAnonymizer
```

Comment:
I have some concerns about the verbosity of the plans generated by this anonymizer and the fact that generated expressions are no longer syntactically valid.

How about anonymizing by canonicalizing all expressions/symbols/literals via a lookup in a map? If an entry is new, synthesize a new concise expression/symbol/literal and add it to the map.

Reply:
The map is just in case there is *not* a one-to-one mapping, so that the same symbol, literal, etc. get anonymized to the same value. The map would just be a "translation table" that's built as new entities are encountered.

So, for instance, for an expression like: `age > minimum_age AND age < maximum_age`, it might do:

1. To anonymize the first term (`age > minimum_age`):
    1. Lookup `age` in translation table. Since it's not found, create a new entry, say `a` and map `age -> a`.
    2. Lookup up `minumum_age` in translation table. Since it's not found, create a new entry, say `b` and map `minimum_age -> b`.
    3. The result would be `a > b`.
2. To anonymize the second term (`age < maximum_age`):
    1. Lookup `age` in translation table, which yields `a`.
    2. Lookup up `maximum_age` in the translation table. Since it's not found, create a new entry, say `c` and map `maximum_age -> c`.
    3. The result would be `a < c`.
3. The final result would be `a > b AND a <  c`.

Reply:
The reason for using hash codes was so that we could say with reasonable confidence (giving some leeway for hash collisions) that if the anonymised plan is the same for 2 queries then it's the same query repeated. It could also be useful for detecting repeated sub-plans across queries.
If we build a translation table on the fly instead of using hash code, we would have to persist that table across queries to allow such an analysis, but then the translation table could increase coordinator memory usage and it would anyway be lost on coordinator restart.
One thought was that we could use hash codes for plans sent to event listener, where readability is less important, and use the translation table for EXPLAIN ANONYMIZED where human readability is a more important goal. We would still need to think about whether the translation table is per-query or persisted across queries though.

Reply:
> It could also be useful for detecting repeated sub-plans across queries.

That's not a goal for this feature. In fact, it could be argued that we need to salt on a per query basis to avoid attempts at de-anonymizing by gathering a big enough sample of queries from the event log.

> If we build a translation table on the fly instead of using hash code, we would have to persist that table across queries to allow such an analysis, but then the translation table could increase coordinator memory usage and it would anyway be lost on coordinator restart.

It's not necessary. The translation would be deterministic if the query/plan structure doesn't change.

> One thought was that we could use hash codes for plans sent to event listener, where readability is less important,

It is just as important. Often times, those plans are going to be displayed in UIs for users to see their query history.

Reply:
We explicitly left the text distributed plan available in the event listener for query history in UIs. The anonymised plans in the event listener are meant for deriving insights about the workload, not for display in UIs. A user should be allowed to see their query/plan without any anonymisation.

Reply:
> A user should be allowed to see their query/plan without any anonymisation.

It depends on which user and where the data is being collected and displayed. Some UIs need to show query plans to users but might not be able to even keep the original plans due to sensitivities around PII.

Reply:
> The translation would be deterministic if the query/plan structure doesn't change.

@martint do you have any specific translation mechanism in mind? I was thinking about a simple counter-based mechanism but it'll lead to all plans having similar symbol names like symbol_0. This method will also make it hard to drive insights. For example, if we want to find the repeated queries/sub queries.

Reply:
Another way to resolve this issue, we could have different types of anonymizers behind some config. For example, `HashingAnonymizer` and `CounterAnonymizer`

WDYT?

Reply:
IMO anonymization of literals should also be consistent across queries, so that we could identify that particular predicate is super common for a given table. It makes difference, if we see 1.000.000 occurrences of different literals vs 1.000.000 of same small subset of literals

BTW: @martint what problem are we trying to solve with counters?

Reply:
> I have some concerns about the verbosity of the plans generated by this anonymizer and the fact that generated expressions are no longer syntactically valid.

`Anonymized` expressions are no longer expressions. They are just strings:
 ```
String anonymize(Expression expression);
```
They are not neccecerly parseable.